arrow/arrow.py

```diff
@@ -9,84 +9,90 @@ def choose(n, k):
-    reactants = np.where(reaction < 0)
+def propensity(stoichiometry, state, form):
+    reactants = np.where(stoichiometry < 0)
     terms = [
-        form(state[reactant], -reaction[reactant])
+        form(state[reactant], -stoichiometry[reactant])
         for reactant in reactants[0]]
 
     return np.array(terms).prod()
 
-
-def step(reactions, rates, state, forms, propensities=[], update_reactions=()):
+def step(stoichiometry, rates, state, forms, propensities=[], update_reactions=()):
     if len(update_reactions):
         for update in update_reactions:
-            reaction = reactions[update]
+            reaction_stoichoiometry = stoichiometry[:, update]
             form = forms if callable(forms) else forms[update]
-            propensities[update] = propensity(reaction, state, form)
+            propensities[update] = propensity(reaction_stoichoiometry, state, form)
     else:
-        propensities = [
-            propensity(reaction, state, forms if callable(forms) else forms[index])
-            for index, reaction in enumerate(reactions)]
+        propensities = np.array([
+            propensity(reaction_stoichoiometry, state, forms if callable(forms) else forms[index])
+            for index, reaction_stoichoiometry in enumerate(stoichiometry.T)])
 
     distribution = (rates * propensities)
     total = distribution.sum()
-    if not total:
-        return (state, 0, -1, propensities)
-
-    dt = np.random.exponential(1 / total)
-    random = np.random.uniform(0, 1) * total
-        
-    progress = 0
-    for choice, interval in enumerate(distribution):
-        progress += interval
-        if random <= progress:
-            break
 
-    reaction = reactions[choice]
-    outcome = state + reaction
+    if total == 0:
+        dt = 0
+        outcome = state
+        choice = -1
+
+    else:
+        dt = np.random.exponential(1 / total)
+        random = np.random.uniform(0, 1) * total
 
-    return outcome, dt, choice, propensities
+        progress = 0
+        for choice, interval in enumerate(distribution):
+            progress += interval
+            if random <= progress:
+                break
 
+        reaction_stoichoiometry = stoichiometry[:, choice]
+        outcome = state + reaction_stoichoiometry
 
-def evolve(reactions, rates, state, duration, forms=choose):
-    time = 0
-    history = [state]
-    steps = [0]
+    return dt, outcome, choice, propensities
+
+
+def evolve(stoichiometry, rates, state, duration, forms=choose):
+    t = 0
+    time = [0]
+    counts = [state]
     propensities = []
     update_reactions = []
 
     while True:
-        state, dt, choice, propensities = step(
-            reactions,
+        dt, state, choice, propensities = step(
+            stoichiometry,
             rates,
             state,
             forms,
             propensities,
             update_reactions)
 
-        time += dt
-        if not dt or time > duration:
+        t += dt
```

Comment:
Not sure about this substitution of `time` for `t`.... in general I try to avoid single letter variable names or abbreviations in general (though `dt` is established enough to evade this tendency).

Reply:
Right, I was a bit annoyed by this too.  I was trying to keep the internal and external meaning of `time` consistent and also liked the parallelism of `t` and `dt`.  Will think on this.

Reply:
Yeah I can see maybe `time` is a misnomer, as it does not represent the entirety of time, but rather a time point. I can also see using `now`.